File: abeln-hb-lattice-2e567b22ecd1/source/Native.cpp
#include "Native.hpp"







Native::Native(Lattice *l){
  cout<<"creating new native"<<endl;
  contacts = new  bool  *** [l->nChains];
  for(int i=0;i<l->nChains;i++){
    contacts[i] = new bool ** [l->nChains];
    int length_i =l->chains[i]->N;
    for(int j=0;j<l->nChains;j++){
      int length_j =l->chains[j]->N;
      contacts[i][j] = new  bool* [length_i];
      for(int k=0;k<length_i;k++){
	contacts[i][j][k]=new bool[length_j];
	for(int l=0;l<length_j;l++){
	  contacts[i][j][k][l]=false;
	}
      }
    }
  }
}




void Native::setNative(Lattice * l){
  cout<<"setting native"<<endl;
  //TO DO should check if not out of bounds
  int totalC=0;
  for(int nc=0;nc<l->nChains;nc++){
    for(int n=0;n<l->chains[nc]->N;n++){
      Residue * res = l->chains[nc]->residues[n];
      for(int k=0;k<6;k++){
	Pos posNB =local[k]+ res->pos;
	posNB.periodicBoundary();
	Residue * resNB = l->getResidue(posNB);
	if(resNB!=NULL){
	  if(resNB->chainNum !=  res->chainNum || abs(resNB->n - res->n) !=1){
	    contacts[res->chainNum][resNB->chainNum][res->n][resNB->n]=true;
	    contacts[resNB->chainNum][res->chainNum][resNB->n][res->n]=true;
	    totalC++;
	  }
	}
      }
    }
  }
  totalC =totalC/ 2;
  totCnat=totalC;
  cout<<"TOTAL NATIVE CONTACTS:" <<totCnat<<endl;
  //l->stats.getLatticeStats(l); //should be set after reading in
  nativeEnergy = l->stats.getEtot();
  cout<<"NATIVE INTERNAL ENERGY:" <<nativeEnergy<<endl;
}
 

File: nelman-kmcLATTICEMODEL/source/Native.py
from typing import List, Dict, Tuple
import numpy as np

class Native:
    def __init__(self, lattice):
        print("creating new native")
        self.contacts = np.zeros(
            (lattice.nChains, lattice.nChains, 
             max(c.N for c in lattice.chains[:lattice.nChains]),
             max(c.N for c in lattice.chains[:lattice.nChains])),
            dtype=bool
        )
    
    def setNative(self, lattice):
        print("setting native")
        total_c = 0
        
        for nc in range(lattice.nChains):
            for n in range(lattice.chains[nc].N):
                res = lattice.chains[nc].residues[n]
                
                for k in range(6):
                    pos_nb = Pos.local[k] + res.pos
                    pos_nb.periodicBoundary()
                    res_nb = lattice.getResidue(pos_nb)
                    
                    if res_nb is not None:
                        if res_nb.chainNum != res.chainNum or abs(res_nb.n - res.n) != 1:
                            self.contacts[res.chainNum][res_nb.chainNum][res.n][res_nb.n] = True
                            self.contacts[res_nb.chainNum][res.chainNum][res_nb.n][res.n] = True
                            total_c += 1
        
        self.tot_c_nat = total_c // 2
        print(f"TOTAL NATIVE CONTACTS: {self.tot_c_nat}")
        self.native_energy = lattice.stats.getEtot()
        print(f"NATIVE INTERNAL ENERGY: {self.native_energy}")
    
    def isNativeContact(self, chain_a: int, chain_b: int, res_a: int, res_b: int) -> bool:
        return self.contacts[chain_a][chain_b][res_a][res_b]
    
    def isBelowNativeEnergy(self, e_tot: int) -> bool:
        return e_tot < self.native_energy
    
    def getNativeEnergy(self) -> int:
        return self.native_energy
    
    def hasNativeStructure(self, n_tot: int) -> bool:
        return n_tot == self.tot_c_nat
